pallets/order-book/src/tests.rs

```diff
@@ -380,6 +389,325 @@ fn fill_order_full_works() {
 	});
 }
 
+mod fill_order_partial {
+	use super::*;
+
+	#[test]
+	fn fill_order_partial_works() {
+		for fulfillment_ratio in 1..100 {
+			new_test_ext().execute_with(|| {
+				let buy_amount = 100 * CURRENCY_AUSD_DECIMALS;
+				let min_fulfillment_amount = 1 * CURRENCY_AUSD_DECIMALS;
+				let sell_ratio = FixedU128::checked_from_rational(3u32, 2u32).unwrap();
+
+				assert_ok!(OrderBook::place_order(
+					ACCOUNT_0,
+					DEV_AUSD_CURRENCY_ID,
+					DEV_USDT_CURRENCY_ID,
+					buy_amount,
+					sell_ratio,
+					min_fulfillment_amount,
+				));
+
+				let (order_id, order) = get_account_orders(ACCOUNT_0).unwrap()[0];
+
+				let fulfillment_ratio = Perquintill::from_percent(fulfillment_ratio);
+				let partial_buy_amount = fulfillment_ratio.mul_floor(buy_amount);
+
+				assert_ok!(OrderBook::fill_order_partial(
+					RuntimeOrigin::signed(ACCOUNT_1),
+					order_id,
+					partial_buy_amount,
+				));
+
+				assert_eq!(
+					AssetPairOrders::<Runtime>::get(DEV_AUSD_CURRENCY_ID, DEV_USDT_CURRENCY_ID),
+					vec![order_id]
+				);
+
+				let expected_sell_amount = OrderBook::convert_with_ratio(
+					order.asset_in_id,
+					order.asset_out_id,
+					order.max_sell_rate,
+					partial_buy_amount,
+				)
+				.unwrap();
+
+				let remaining_buy_amount = buy_amount - partial_buy_amount;
+
+				assert_eq!(
+					System::events()[2].event,
+					RuntimeEvent::OrmlTokens(orml_tokens::Event::Unreserved {
+						currency_id: DEV_USDT_CURRENCY_ID,
+						who: ACCOUNT_0,
+						amount: expected_sell_amount
+					})
+				);
+				assert_eq!(
+					System::events()[3].event,
+					RuntimeEvent::OrderBook(Event::OrderUpdated {
+						order_id,
+						account: order.placing_account,
+						buy_amount: remaining_buy_amount,
+						sell_rate_limit: order.max_sell_rate,
+						min_fulfillment_amount: order.min_fulfillment_amount,
+					})
+				);
+				assert_eq!(
+					System::events()[4].event,
+					RuntimeEvent::OrmlTokens(orml_tokens::Event::Transfer {
+						currency_id: DEV_AUSD_CURRENCY_ID,
+						to: ACCOUNT_0,
+						from: ACCOUNT_1,
+						amount: partial_buy_amount
+					})
+				);
+				assert_eq!(
+					System::events()[5].event,
+					RuntimeEvent::OrmlTokens(orml_tokens::Event::Transfer {
+						currency_id: DEV_USDT_CURRENCY_ID,
+						to: ACCOUNT_1,
+						from: ACCOUNT_0,
+						amount: expected_sell_amount
+					})
+				);
+				assert_eq!(
+					System::events()[6].event,
+					RuntimeEvent::OrderBook(Event::OrderFulfillment {
+						order_id,
+						placing_account: order.placing_account,
+						fulfilling_account: ACCOUNT_1,
+						partial_fulfillment: true,
```

Comment:
Yes, but it doesn't get to 100 in this test. The 100 percent case is handled separately given the different expected events.

Reply:
Ah dumb me forgetting that `0..100 != 0..=100` in Rust 🙈

Reply:
Nah, that's on me since I did not find a proper way to have a separate test for each percentage using a macro. If we get a failure here now, one would have to debug and figure out which percentage is problematic, which is awkward to say the least.

Maybe @lemunozm can help ^^

Reply:
IMO it's the current test is more than sane. Theoretically, you could add a debug message to the assertion(s) to populate the failure `fulfillment_ratio`.